src/theory/datatypes/datatypes_rewriter.cpp

```diff
@@ -148,6 +148,101 @@ RewriteResponse DatatypesRewriter::postRewrite(TNode in)
       return RewriteResponse(REWRITE_AGAIN_FULL, ret);
     }
   }
+  else if (k == MATCH)
+  {
+    Trace("dt-rewrite-match") << "Rewrite match: " << in << std::endl;
+    Node h = in[0];
+    std::vector<Node> cases;
+    std::vector<Node> rets;
+    TypeNode t = h.getType();
+    const Datatype& dt = t.getDatatype();
+    for (unsigned k = 1, nchild = in.getNumChildren(); k < nchild; k++)
+    {
+      Node c = in[k];
+      Node cons;
+      Kind ck = c.getKind();
+      if (ck == MATCH_CASE)
+      {
+        Assert(c[0].getKind() == APPLY_CONSTRUCTOR);
+        cons = c[0].getOperator();
+      }
+      else if (ck == MATCH_BIND_CASE)
+      {
+        if (c[1].getKind() == APPLY_CONSTRUCTOR)
+        {
+          cons = c[1].getOperator();
+        }
+      }
+      else
+      {
+        cases.push_back(Node::null());
+        rets.push_back(c);
+        continue;
+      }
```

Comment:
Right, I'll make this an assertion.

Reply:
It may be slightly better to make it `Unreachable()`?

Reply:
What is the difference?

Reply:
They act in a very similar fashion but `Unreachable()` indicates (at least for me) that we don't know _how_ it would be possible that this code gets executed. So for me, it is almost a stronger indicator than an assertion. However, this is just a detail and I am fine either way.